docs(*): remove usage of global grunt-cli

frederikprijck opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Docs update


**What is the current behavior? (You can also link to an open issue here)**
Currently, the contributing docs tell people to globally install grunt and build angularjs using the `grunt` cli command.


**What is the new behavior (if this is a feature change)?**
This commit ensures grunt is not required globally by updating all the commands to use `yarn grunt` instead of `grunt`.


**Does this PR introduce a breaking change?**
No


**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

frederikprijck commented:
Personally, I prefer to use `scripts`. This would result in commands such as `yarn package` instead of `yarn grunt package`. This can still be modified in the current PR if wanted, but I'd like to discuss it prior to changing it.

Narretz commented:
How about we add a note that you don't have to install `grunt-cli` globally if you prefix the grunt commands with `yarn`?

frederikprijck commented:
But what's the point of using the global grunt cli by default? Using global modules for these kind of things doesn't have any benefits, does it?

petebacondarwin commented:
My reading of this PR is that it is not using a global grunt install anymore.

Narretz commented:
> Using global modules for these kind of things doesn't have any benefits, does it?

It does have the benefit that you don't have to prefix each command with `yarn `. That's 5 keystrokes. 

> Do you mean to undo the changes of this commit and only add the note?

That's right. Another problem is that this is not the only place where these instructions are. There's also https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md, https://github.com/angular/angular.js/blob/master/README.md, even on the angularjs.org repo: https://github.com/angular/angularjs.org

gkalpak commented:
> Personally, I prefer to use scripts. This would result in commands such as yarn package instead of yarn grunt package.

For better or worse, we are using `grunt` as our task runner. At this point, I don't see much benefit in implementing "aliases" for all grunt commands (and having to keep the yarn scripts up-to-date with the grunt commands in case we add/remove any).

##
> How about we add a note that you don't have to install grunt-cli globally if you prefix the grunt commands with yarn?

I would go the other way and document that you can drop the `yarn` part in `yarn grunt ...` if you install `grunt` globally (but keep the uing the local version "by default").

👍 for updating all places which refer to `grunt` commands :smiley:

frederikprijck commented:
> At this point, I don't see much benefit in implementing "aliases" for all grunt commands

Ye makes sense. 

Using scripts does hide the actual implementation from the usage of the commands (same cli command, whether you're using grunt, gulp, webpack, pure node, ...).
As AngularJS is using both Grunt and Gulp (atleast, they are both installed), one could run grunt and gulp without the user/developer needs to know which one he is running by using scripts.

But yes, the downside is you need to maintain the scripts.

Narretz commented:
If we document `yarn grunt ...` as default, with the option to install `grunt-cli` globally, then that's fine with me

Narretz commented:
This PR must / can be updated now. There's now things to change in the DEVELOPERS.md, too.

frederikprijck committed:
docs(*): remove usage of global grunt-cli

Previously, the `DEVELOPERS.md` and `CONTRIBUTING.md` files
refered to global `grunt-cli` by default.

This commit ensures the local `grunt-cli` is used by default
and mentiones the possibility to still use the global `grunt-cli`.

frederikprijck commented:
Alright, I've moved the changes to the new files. I didn't add the message regarding global gunt yet, will do soon.

I was wondering how this ended up in here and I wanted to check if this is realy neccesary ?

> Note: If you're using Linux, and yarn install fails with the message 'Please try running this command again as root/Administrator.', you may need to globally install grunt and bower:

I've had this problem on my mac for a while (prior to my npm permissions where set correctly).
Imho this should not result in global installation of the npm packages, instead there could be two solutions:

- sudo the local commands (but we don't want that)
- fix permission issues.

Narretz commented:
I always go to the git blame to find why something was added. Leads to result in 50% of the cases. ;)
This was originally added for `npm` and then simply changed to `yarn`. https://github.com/angular/angular.js/commit/1dedcdf2bc7f6adaa3977d12d68de1ba0a87150c#diff-acca20ff7c21d5e56c51120c19837a7fL81  I'm okay with adding a link to yarn docs for permission issues.

Narretz commented on DEVELOPERS.md:
````diff
@@ -70,20 +63,13 @@ git remote add upstream "https://github.com/angular/angular.js.git"
-sudo yarn global add grunt-cli
-sudo yarn global add bower
-```
+**Note:** If you're using Linux, and `yarn install` fails with the message 'Please try running this command again as root/Administrator.', you may need to fix your [npm permissions](https://docs.npmjs.com/getting-started/fixing-npm-permissions).
````
Why npm when we are using yarn?

frederikprijck replied:
That's a good question, it's a common issue that permissions are set wrong on osx and Linux for npm, but I'll check how that works for yarn!

frederikprijck replied:
@Narretz I had a chat offline with @gkalpak and we think it's probably a good idea to just leave this part out and reconsider if there are issues with this. What do you think?

Narretz replied:
Fine with me

frederikprijck commented:
I've added the mentioning of installing grunt globally:

[Grunt](http://gruntjs.com): We use Grunt as our build system. We're using it as a local dependencies but if you prefer to use the command-line tool
 +  globally, you can install it using yarn:
  
```shell
    yarn global add grunt-cli
```

I'm still a bit unsure what the benefit is for adding this to the docs, as it realy is not recommended (it's even recommended to avoid it) to install it globally. If it's not recommended, why guide anyone to adding it globally ?

Narretz commented:
Because then you can `grunt ...` instead of `yarn grunt ...` At the moment, the message isn't very helpful if you don't mentiond that you can skip the `yarn` part in that case

frederikprijck commented:
Let me add that part aswell.

> Because then you can grunt ... instead of yarn grunt ...

Even tho thats true, the issues global NPM modules can cause are worse.
Imagine contributing to several projects, each using a seperate, globally, grunt version ...

I do agree that if that's not the case, installing it globally results in "shorter" commands. But promoting the global installation is odd.

frederikprijck commented:
> The yarn docs also say there are use cases for global adding

Ofcourse there are. Every none-project dependency should be globally (`yarn, `yeoman`, `angular/cli`, ...).
Every project dependency should be in the project in such a way that the project is self-contained, resulting in less hassle to setup a project.

Narretz commented on DEVELOPERS.md:
```diff
@@ -34,13 +34,15 @@ machine:
 
-* [Grunt](http://gruntjs.com): We use Grunt as our build system. Install the grunt command-line tool
-  globally with:
+* [Grunt](http://gruntjs.com): We use Grunt as our build system. We're using it as a local dependencies but if you prefer to use the command-line tool
```
Sorry to bother you so long with this :D but I feel like we can be more precise here. How about 

> We're using it as a local dependency, but you can also add the grunt command-line tool globally, which allows you to leave out the `yarn` prefix for all our grunt commands:

Then you can remove the note below (which makes it sound like you can only use global or local grunt, but you can still use yarn grunt ... if you want, even with global grunt-cli).

frederikprijck replied:
No need to feel sorry, I appreciate the feedback! I do think it reads better the way u propose, let me update it!

Narretz replied:
I will merge  when you've updated this part 👍

frederikprijck replied:
Done 😄
